third_party/pcsc-lite/naclport/server_clients_management/src/admin_policy_getter.cc

```diff
@@ -0,0 +1,98 @@
+// Copyright 2022 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "admin_policy_getter.h"
+
+#include <condition_variable>
+#include <string>
+
+#include <google_smart_card_common/optional.h>
+#include <google_smart_card_common/value_conversion.h>
+#include <google_smart_card_common/value_debug_dumping.h>
+
+namespace google_smart_card {
+
+namespace {
+
+constexpr char kUpdateAdminPolicyMessageType[] = "update_admin_policy";
+
+}  // namespace
+
+template <>
+StructValueDescriptor<AdminPolicy>::Description
+StructValueDescriptor<AdminPolicy>::GetDescription() {
+  // Note: Strings passed to WithField() below must match the keys in
+  // //third_party/pcsc-lite/naclport/server_clients_management/src/client-handler.js.
+  return Describe("AdminPolicy")
+      .WithField(&AdminPolicy::force_allowed_client_app_ids,
+                 "force_allowed_client_app_ids")
+      .WithField(&AdminPolicy::scard_disconnect_fallback_client_app_ids,
+                 "scard_disconnect_fallback_client_app_ids")
+      .PermitUnknownFields();
+}
+
+AdminPolicyGetter::AdminPolicyGetter() = default;
+
+AdminPolicyGetter::~AdminPolicyGetter() = default;
+
+optional<AdminPolicy> AdminPolicyGetter::WaitAndGet() {
+  // Wait until the policy is received for the first time
+  std::unique_lock<std::mutex> lock(mutex_);
+  condition_variable_.wait(lock, [this]() {
+    return shutting_down_ || admin_policy_received_;
+  });
+  if (shutting_down_)
+    return {};
+  return admin_policy_;
+}
+
+void AdminPolicyGetter::ShutDown() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  shutting_down_ = true;
+  condition_variable_.notify_all();
+}
+
+std::string AdminPolicyGetter::GetListenedMessageType() const {
+  return kUpdateAdminPolicyMessageType;
+}
+
+bool AdminPolicyGetter::OnTypedMessageReceived(Value data) {
+  AdminPolicy message_data;
+  std::string error_message;
+  if (!ConvertFromValue(std::move(data), &message_data, &error_message)) {
+    GOOGLE_SMART_CARD_LOG_WARNING << "Failed to parse admin policy message: "
+                                  << error_message;
+    return false;
```

Comment:
IIRC this will lead to a crash. I'm not sure it's the appropriate action, given that the format of the policy is received from Chrome, and we shouldn't blindly rely on the latter to do all the checks correctly.

Let's log the error, but still avoid the crash. Also makes sense to pretend that we received an (empty) policy value, so that `WaitAndGet()` callers are unblocked.